### PR TITLE
MLDB-1345 evaluate HAVING expressions

### PR DESCRIPTION
**Pull request by steadtler**
The HAVING was already part of the SQL statement and parsed and everything and even in the doc, so I only had to correctly evaluate it in the groupby query.

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
can you remove the 'manual' on the test case so we can see that it passes please? :)

File: core/dataset.cc [hunk 1]
```diff
@@ -551,8 +551,12 @@ queryStructured(const SelectExpression & select,
     std::mutex lock;
     std::vector<MatrixNamedRow> output;
 
+    if (!having.isConstantTrue() && groupBy.clauses.empty())
+        throw HttpReturnException(400, "HAVING expression requires a GROUP BY expression");
```
Comment (jeremybarnes) [inline core/dataset.cc]: Is that true?  What about something with an aggregator but no group by... is HAVING illegal there?

File: sql/sql_expression.cc [hunk 2]
```diff
@@ -1740,6 +1747,46 @@ unimp(std::shared_ptr<SqlExpression> lhs,
...
+    std::vector<std::shared_ptr<SqlExpression> > output;
+    std::vector<std::shared_ptr<SqlExpression> > children = getChildren();
+
+    int index = 0;
+    while(index < children.size()) {
+        auto child = children[index];
+
+        bool foundAggregator = false;
+        if (child->getType() == "function") {
+            const FunctionCallWrapper * function = dynamic_cast<const FunctionCallWrapper *>(child.get());
```
Comment (jeremybarnes) [inline sql/sql_expression.cc]: Why is it called FunctionCallWrapper and not FunctionCallExpression?

**Comment by jeremybarnes** [comment]
Looks OK to me.  +1 once Nic's comments are addressed.

File: core/dataset.cc (hunk 1 above)
Comment (steadtler) [inline core/dataset.cc]: According to the "internet" a having absolutely requires a group by. If you have aggregators but no group by, you get only 1 row anyway, so filtering that row has little use. Nic can overrule me here though...

File: sql/sql_expression.cc (hunk 2 above)
Comment (steadtler) [inline sql/sql_expression.cc]: It was originally a wrapper around builtin function calls and user function calls back when the syntax and usage was more different. We could rename it.

**Pull request merged**